"Working with LINQ" tutorial improvements

jmarkman opened the pull request:
## Summary

I don't know what Microsoft has in store for the future in regards to improving the docs experience overall, but I took @mairaw's 

> make it easier to follow along

part of her comment in the ticket to heart for this pull request. My goals for the PR were to:

- **make sure that following the docs article had an orderly flow to it from start to finish**

Example: The doc starts with the LINQ query that builds the deck of cards. Not even inside a `Main()` method, but just there in a snippet. Meanwhile, the `using` statements that the reader should be aware of from the start are at the very end of the **Creating our Data Set** section. Sure, it's easy to argue that C# developers should know their syntax A-B-Cs as a tenant of programming, but I don't believe that such an argument is justification to just place information willy-nilly in a section and believe that that the reader will get to it eventually™.

- **add overall clarity to the writing**

Example: adding hyperlinks to other docs for concepts that are mentioned yet not talked about, like iterator methods and extension methods. I also added comments in code snippets denoting what file said code snippets would be in.

- **add context**

Example:  when `Take()` and `Skip()` are first introduced, the only explanation for their relevance to our program is a singular code block. The reader only finds out that they're going to be used _in_ the demo is _after_ they write their first extension method. To fix this, I explicitly state where these lines will go, include the entire `Main()` method as a code snippet, place `Take()` and `Skip()` in that block, and add a brief comment for the mathematical reasoning

- **simplify**

Example: For the most part, I thought that the sentence content of the doc was fairly simple. However, after following the doc from start to finish, I realized that the entire **Preparing for New Features** section is completely irrelevant to LINQ in its entirety. Yes, it's showing good practices, but there is not one line in that section that talks about LINQ at all. It's very distracting from the content of the doc and leads readers who skim before reading to believe that the doc page is much more complicated than it really is.

After working with this docs page, I ended up developing a lot of misgivings about how much of a tutorial about a deck of cards could provide. Yes, you can generate a deck of cards as shown by the doc, but something a bit more "real" might be in the interest of this doc page, like working with data source from a JSON or XML file, or a database (though not necessarily going through the steps to get the data _from_ those sources). Something that requires more than just three LINQ statements; show readers the power of statements like `SelectMany()`, `GroupBy()`, and so on alongside the documentation for those statements.

Looking forward to feedback.

Fixes #7829

Commit:
updates
- reordered Creating the Data Set
- made reading and working through Creating the Data Set easier
- included translation from query syntax to method syntax
- linked to iterator methods section in iterators.md

Commit:
further changes
- started reorganization of Mainpulating the Order
- provided a link to the Extension Methods page
- explicitly stated to create a new file for extension methods
- gave context to where Take() and Skip() should go

Commit:
further updates
- added "comments" to signal what code snippets belonged to which .cs files
- added context to the first code snippet for Comparisons
- clarified section on terminal idioms
- clarified final section detailing results before Optimizations

jmarkman committed:
more clarity changes
- added link to Deferred Execution page
- denoted which code blocks represented inner and outer shuffles
- gave single-sentence explanationon lazy evaluation
- made the "this is a designed example" section more clear

jmarkman committed:
one last update before pr
- Axed entirety of Preparing for New Features
- Focused and trimmed the conclusion

pkulikov commented on docs/csharp/tutorials/working-with-linq.md:
````diff
@@ -80,64 +77,97 @@ static IEnumerable<string> Ranks()
     yield return "ace";
 }
 ```
+Place these underneath the `Main` method in your `Program.cs` file. These two methods both utilize the `yield return` syntax to produce a sequence as they run. The compiler builds an object that implements `IEnumerable<T>` and generates the sequence of strings as they are requested.
````
I like that you've added links to related docs (iterator, extension methods). You even can go further and add links to API docs as well (so called, xref-links). For example, replace `IEnumerable<T>` with `System.Collections.Generic.IEnumerable%601` (without backticks) to generate the link to [`IEnumerable<T>`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.ienumerable-1?view=netframework-4.7.2)

jmarkman replied:
Sure, I can definitely make this happen

pkulikov commented on docs/csharp/tutorials/working-with-linq.md:
```diff
@@ -80,64 +77,97 @@ static IEnumerable<string> Ranks()
+
+You can see the addition of the `this` modifier on the first argument to the method. That means you call the method as though it were a member method of the type of the first argument. This method declaration also follows a standard idiom where the input and output types are `IEnumerable<T>`. That practice enables LINQ methods to be chained together to perform more complex queries.
+
+Naturally, since we split the deck into halves, we'll need to join those halves together. In code, this means we'll be enumerating both of the sequences we acquired through `Take()` and `Skip()` at once, *interleaving* the elements, and creating one object: our now-shuffled deck of cards. Writing a LINQ method that works with two sequences requires that you understand how `IEnumerable` works.
```
At the end of the paragraph: `IEnumerable` -> `IEnumerable<T>`, unless you really mean the non-generic `IEnumerable` interface.

jmarkman replied:
Yeah, that's a good point. I believe I left it as-is because the original doc also had it as a plain `IEnumerable`. I'll make sure that it's the generic interface here and see if I forgot to include it elsewhere.

pkulikov commented on docs/csharp/tutorials/working-with-linq.md:
```diff
@@ -80,64 +77,97 @@ static IEnumerable<string> Ranks()
+
+Naturally, since we split the deck into halves, we'll need to join those halves together. In code, this means we'll be enumerating both of the sequences we acquired through `Take()` and `Skip()` at once, *interleaving* the elements, and creating one object: our now-shuffled deck of cards. Writing a LINQ method that works with two sequences requires that you understand how `IEnumerable` works.
+
+The `IEnumerable` interface has one method: `GetEnumerator()`. The object returned by `GetEnumerator()` has a method to move to the next element, and a property that retrieves the current element in the sequence. You will use those two members to enumerate the collection and return the elements. This Interleave method will be an iterator method, so instead of building a collection and returning t ...
```
First sentence: `IEnumerable` -> `IEnumerable<T>` and consider xref-link to [GetEnumerator method](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.ienumerable-1.getenumerator?view=netframework-4.7.2)

pkulikov commented on docs/csharp/tutorials/working-with-linq.md:
```diff
@@ -270,62 +313,22 @@ public static void Main(string[] args)
 Notice that you don't log every time you access a query. You log only when you create the original query. The program still takes a long time to run, but now you can see why. If you run out of patience running the inner shuffle with logging turned on, switch back to the outer shuffle. You'll still see the lazy evaluation effects. In one run, it executes 2592 queries, including all the value and s ...
 
-There is an easy way to update this program to avoid all those executions. There are LINQ methods `ToArray()` and `ToList()` that cause the query to run, and store the results in an array or a list, respectively. You use these methods to cache the data results of a query rather than execute the source query again.  Append the queries that generate the card decks with a call to `ToArray()` and run ...
+We can improve the performance of our code here to reduce the number of executions we make. A simple fix we can make is to *cache* the results of our original LINQ query that constructs the deck of cards. Currently, we're executing our queries again and again every time our do-while loop goes through an iteration, re-constructing our deck of cards and resshuffling it every time. To cache our deck ...
```
Again, consider links to `ToArray` and `ToList` API docs.

pkulikov commented on docs/csharp/tutorials/working-with-linq.md:
```diff
@@ -270,62 +313,22 @@ public static void Main(string[] args)
+* lazy and eager evaluation in regards to LINQ queries and the implications they might have on query performance
 
-You learned a bit about one magician's technique. Magicians use the faro shuffle because they can control where every card moves in the deck. In some tricks, the magician has an audience member place a card on top of the deck, and shuffles a few times, knowing where that card goes. Other illusions require the deck set a certain way. A magician will set the deck prior to performing the trick. Then ...
+Aside from LINQ, we learned a bit about a technique magicians use for card tricks. Magicians use the Faro shuffle because they can control where every card moves in the deck. Now that you know, don't spoil it for everyone else!
```
Docs contain many more examples of LINQ usage. I think they should be referenced from this tutorial in the **See also** section. Articles to reference might be:
https://docs.microsoft.com/en-us/dotnet/csharp/linq/
https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/linq/

jmarkman replied:
I have a few pages in mind, but how would I link those two pages? I don't think I can reference folders using traditional markdown syntax, i.e.:

`[Language Integrated Query (LINQ)](../the-linq-page)`

Where "the-linq-page" = https://docs.microsoft.com/en-us/dotnet/csharp/linq/
Would this just be a use case for hyperlinks?

pkulikov replied:
@jmarkman cool, thanks!

> I'd almost worry that docs pages would begin to circularly reference each other.

I don't think it's a big issue. Sometimes, I need to read the same page twice, with a break between reading sessions, to get full understanding of the topic. Circular referencing fits that workflow 😄

pkulikov reviewed (commented):
@jmarkman I've left several suggestions to add even more references. And please update the `ms.date` field at the line 4 :).

Commit:
updated with feedback
- updated ms.date
- added links to API docs with xref
- added links to more information about LINQ

jmarkman commented on docs/csharp/tutorials/working-with-linq.md:
````diff
@@ -80,64 +77,97 @@ static IEnumerable<string> Ranks()
+}
+```
+
+The multiple `from` clauses produce a <xref:System.Linq.Enumerable.SelectMany>, which creates a single sequence from combining each element in the first sequence with each element in the second sequence. The order is important for our purposes. The first element in the first source sequence (Suits) is combined with every element in the second sequence (Values). This produces all thirteen cards of ...
````
Do you think it'd be worth adding to the `SelectMany` xref-links just in case?

pkulikov replied:
@jmarkman 
>Do you think it'd be worth adding to the SelectMany xref-links just in case?

Yes, and also to other method xref-links

Commit:
added "%2A" to xref-links as advised by @pkulikov

BillWagner commented:
@jmarkman Thanks for your patience. I've been a bit swamped, and you made a number of improvements to this article. I want to dedicate the time to review it well. I will carve out time either later today or tomorrow morning to give it the time and focus it deserves.

jmarkman commented:
@BillWagner I know the feeling, there's no rush. I appreciate the consideration.

BillWagner commented on docs/csharp/tutorials/working-with-linq.md:
````diff
@@ -174,49 +205,58 @@ Let's see how many shuffles it takes to set the deck back to its original order.
 ```
 
-Run the sample, and see how the deck rearranges on each shuffle, until it returns to its original configuration after 8 iterations.
+Run the code we've got so far. Notice how the deck rearranges on each shuffle; after 8 shuffles (iterations of the do-while loop), the deck returns to its original configuration it was in when we first created it from our starting LINQ query.
````
instead of the semi-colon, this should be 2 sentences. 

"its original configuration it was in"  -> consider "the original configuration it was in"

BillWagner commented on docs/csharp/tutorials/working-with-linq.md:
```diff
@@ -174,49 +205,58 @@ Let's see how many shuffles it takes to set the deck back to its original order.
 ## Optimizations
 
-The sample you've built so far executes an *out shuffle*, where the top and bottom cards stay the same on each run. Let's make one change, and run an *in shuffle*, where all 52 cards change position. For an in shuffle, you interleave the deck so that the first card in the bottom half becomes the first card in the deck. That means the last card in the top half becomes the bottom card. That's just  ...
+The sample you've built so far executes an *outer shuffle*, where the top and bottom cards stay the same on each run. Let's make one change: we'll use an *inner shuffle* instead, where all 52 cards change position. For an inner shuffle, you interleave the deck so that the first card in the bottom half becomes the first card in the deck. That means the last card in the top half becomes the bottom  ...
```
The terms "in shuffle" and "out shuffle" were correct. (I know, it sounds weird to me too).

BillWagner commented on docs/csharp/tutorials/working-with-linq.md:
```diff
@@ -242,6 +283,7 @@ public static void Main(string[] args)
 
     do
     {
+        // Outer shuffle
```
again, here it should be "in shuffle" and "out shuffle"

BillWagner commented on docs/csharp/tutorials/working-with-linq.md:
```diff
@@ -270,62 +313,31 @@ public static void Main(string[] args)
+* locating areas in our code where our LINQ queries might run into performance issues like degraded speed
+* lazy and eager evaluation in regards to LINQ queries and the implications they might have on query performance
+
+Aside from LINQ, we learned a bit about a technique magicians use for card tricks. Magicians use the Faro shuffle because they can control where every card moves in the deck. Now that you know, don't spoil it for everyone else!
```
nice.  😺

BillWagner reviewed (changes requested):
Overall, this is great work. We really appreciate it @jmarkman.  These updates do make this more approachable.

I did have a few suggestions and style considerations.  The most important is to check for all 1st person references. The overall tutorial will read much better if all the references are 2nd person.

Thanks again, and I'll watch for updates.

jmarkman commented on docs/csharp/tutorials/working-with-linq.md:
```diff
@@ -35,26 +35,23 @@ If you've never used C# before, [this tutorial](console-teleprompter.md) explain
-Let's start by creating a deck of cards. You'll do this using a LINQ query that has two sources (one for the four suits, one for the thirteen values). You'll combine those sources into a 52 card deck. A `Console.WriteLine` statement inside a `foreach` loop displays the cards.
-
-Here's the query:
+Before we begin, make sure that the following lines are at the top of the `Program.cs` file generated by `dotnet new console`:
```
Ah, I read the "voice tone" part of the style guide, but I guess I was stuck in my own ways while writing. I can change it.

BillWagner replied:
😆 I spot those because it was also my style for a long time.

jmarkman commented on docs/csharp/tutorials/working-with-linq.md:
```diff
@@ -35,26 +35,23 @@ If you've never used C# before, [this tutorial](console-teleprompter.md) explain
+If these three lines (`using` statements) aren't at the top of the file, our program will not compile.
 
-Next, you'll need to build the Suits() and Ranks() methods. Let's start with a really simple set of *iterator methods* that generate the sequence as an enumerable of strings:
+Now that we have all of the references we need, let's consider what constitutes a deck of cards. Commonly, a deck of playing cards has four suits, and each suit has thirteen values. Normally, you might consider creating a `Card` class right off the bat and populating a collection of `Card` objects by hand. With LINQ, we can be more concise than the usual way of dealing with creating a deck of car ...
```
Will change as well. My thought process for saying "let's consider" over "consider was that it would be more conversational rather than a textbook-like "consider" (i.e., a calculus textbook that would say "Consider the formula {blah}").

jmarkman commented on docs/csharp/tutorials/working-with-linq.md:
```diff
@@ -80,64 +77,97 @@ static IEnumerable<string> Ranks()
+You can see the addition of the `this` modifier on the first argument to the method. That means you call the method as though it were a member method of the type of the first argument. This method declaration also follows a standard idiom where the input and output types are `IEnumerable<T>`. That practice enables LINQ methods to be chained together to perform more complex queries.
 
-The `IEnumerable` interface has one method: `GetEnumerator()`. The object returned by `GetEnumerator()` has a method to move to the next element, and a property that retrieves the current element in the sequence. You will use those two members to enumerate the collection and return the elements. This Interleave method will be an iterator method, so instead of building a collection and returning t ...
+Naturally, since we split the deck into halves, we'll need to join those halves together. In code, this means we'll be enumerating both of the sequences we acquired through <xref:System.Linq.Enumerable.Take%2A> and <xref:System.Linq.Enumerable.Skip%2A> at once, *interleaving* the elements, and creating one object: our now-shuffled deck of cards. Writing a LINQ method that works with two sequences ...
```
I'm definitely ignorant about the usage of sequence instead of object or collection. I'll change this to sequence since I've seen it in other docs pages already. Is sequence more broad/a catch-all of sorts?

BillWagner replied:
Good question @jmarkman 

I'd prefer "sequence" over "object" here because I'd usually think of "object" as a single item. The "sequence" in LINQ would mean any `IEnumerable<T>`.

jmarkman committed:
updated "working with linq" page based on feedback from @BillWagner
- adjusted doc to use second person as per "voice tone" style guide
- code formatting
- using "sequence" instead of "object" where applicable
- code-fence the word "interleave"
- various sentence adjustments

jmarkman commented:
Did a bunch of these updates later at night, let me know if anything slipped through.

BillWagner commented:
Thanks for updating this, and responding to all the reviews. The article looks great now.

I'll :shipit: today. The updates should show up on the live site in the few days, on our regular publishing cycle.  Thanks again for contributing to docs! We appreciate it. 🎆
